File: platform/polyglot/fetch_http_cache_response/py/fetch_http_cache_response/sdk.py
"""SDK factory functions and convenience methods (Story 4)."""

from __future__ import annotations

import time
from typing import Any

from .client import FetchHttpCacheClient, _generate_cache_key
from .exceptions import FetchCacheConfigError
from .logger import create as create_logger
from .token_manager import create_token_manager
from .types import CacheResponseConfig, FetchResult, SDKConfig

logger = create_logger(__name__)


# ─── SDK Factory Functions ────────────────────────────────────────────────────


def create_fetch_cache_sdk(config: SDKConfig) -> FetchHttpCacheClient:
    """Create a FetchHttpCacheClient from SDKConfig."""
    token_manager = None
    if config.auth is not None:
        token_manager = create_token_manager(config.auth)

    return FetchHttpCacheClient(
        config=config,
        token_manager=token_manager,
    )


def create_fetch_cache_sdk_from_yaml(
    yaml_section: dict[str, Any],
    **overrides: Any,
) -> FetchHttpCacheClient:
    """Create SDK from YAML config section with three-tier resolution."""
    config = SDKConfig.from_env(yaml_config=yaml_section, **overrides)
    return create_fetch_cache_sdk(config)


def create_fetch_cache_sdk_from_env(**overrides: Any) -> FetchHttpCacheClient:
    """Create SDK from environment variables only."""
    config = SDKConfig.from_env(**overrides)
    return create_fetch_cache_sdk(config)


# ─── Convenience Functions ────────────────────────────────────────────────────


async def fetch_cached(
    url: str,
    method: str = "GET",
    headers: dict[str, str] | None = None,
    body: Any = None,
    config: SDKConfig | None = None,
    **overrides: Any,
) -> FetchResult[Any]:
    """One-shot stateless fetch with caching."""
    if config is None:
        config = SDKConfig.from_env(**overrides)

    async with create_fetch_cache_sdk(config) as client:
        return await client.request(method, url, headers=headers, body=body)


async def invalidate_cache(
    key_or_url: str,
    config: SDKConfig | None = None,
    **overrides: Any,
) -> None:
    """Manual cache invalidation by key or URL."""
    if config is None:
        config = SDKConfig.from_env(**overrides)

    # If it looks like a URL, generate the key
    if key_or_url.startswith(("http://", "https://")):
        key = _generate_cache_key(
            method="GET",
            url=key_or_url,
            key_strategy=config.cache.key_strategy,
            key_prefix=config.cache.key_prefix,
        )
    else:
        key = key_or_url

    async with create_fetch_cache_sdk(config) as client:
        await client.invalidate_cache(key)

File: platform/polyglot/fetch_http_cache_response/mjs/src/sdk.ts
/**
 * SDK factory functions and convenience methods (Story 4).
 */

import { FetchHttpCacheClient, generateCacheKey } from "./client.js";
import { resolveConfig } from "./config.js";
import { createLogger } from "./logger.js";
import { createTokenManager } from "./token-manager.js";
import type { FetchResult, ResolvedSDKConfig, SDKConfig } from "./types.js";

const logger = createLogger("sdk");

// ─── SDK Factory Functions ──────────────────────────────────────────────────

export function createFetchCacheSdk(
  config: SDKConfig,
  yamlConfig?: Record<string, unknown> | null,
): FetchHttpCacheClient {
  const resolved = resolveConfig(config, yamlConfig);
  const tokenManager = resolved.auth !== null
    ? createTokenManager(resolved.auth)
    : null;
  return new FetchHttpCacheClient(resolved, null, null, tokenManager);
}

export function createFetchCacheSdkFromYaml(
  yamlSection: Record<string, unknown>,
  overrides?: SDKConfig,
): FetchHttpCacheClient {
  return createFetchCacheSdk(overrides ?? {}, yamlSection);
}

export function createFetchCacheSdkFromEnv(
  overrides?: SDKConfig,
): FetchHttpCacheClient {
  return createFetchCacheSdk(overrides ?? {});
}

// ─── Convenience Functions ──────────────────────────────────────────────────

export async function fetchCached(
  url: string,
  options: {
    method?: string;
    headers?: Record<string, string>;
    body?: unknown;
    config?: SDKConfig;
  } = {},
): Promise<FetchResult> {
  const client = createFetchCacheSdk(options.config ?? {});
  try {
    return await client.request(
      options.method ?? "GET",
      url,
      { headers: options.headers, body: options.body },
    );
  } finally {
    await client.close();
  }
}

export async function invalidateCache(
  keyOrUrl: string,
  config?: SDKConfig,
): Promise<void> {
  const resolved = resolveConfig(config ?? {});

  let key: string;
  if (keyOrUrl.startsWith("http://") || keyOrUrl.startsWith("https://")) {
    key = generateCacheKey(
      "GET", keyOrUrl,
      resolved.cache.keyStrategy,
      resolved.cache.keyPrefix,
    );
  } else {
    key = keyOrUrl;
  }

  const client = createFetchCacheSdk(config ?? {});
  try {
    await client.invalidateCache(key);
  } finally {
    await client.close();
  }
}
